src/mscorlib/shared/System/Buffers/OwnedMemory.cs

```diff
@@ -0,0 +1,65 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    public abstract class OwnedMemory<T> : IDisposable, IRetainable
+    {
+        protected OwnedMemory() { }
+
+        public abstract int Length { get; }
+
+        public abstract Span<T> AsSpan(int index, int length);
+
+        public virtual Span<T> AsSpan()
+        {
+            if (IsDisposed) ThrowHelper.ThrowObjectDisposedException(nameof(OwnedMemory<T>), ExceptionResource.Memory_ThrowIfDisposed);
+            return AsSpan(0, Length);
+        }
+
+        public Memory<T> Memory
+        {
+            get 
+            {
+                if (IsDisposed) ThrowHelper.ThrowObjectDisposedException(nameof(OwnedMemory<T>), ExceptionResource.Memory_ThrowIfDisposed);
+                return new Memory<T>(this, 0, Length);
+            }
+        }
+
+        public ReadOnlyMemory<T> ReadOnlyMemory
+        {
+            get 
+            {
+                if (IsDisposed) ThrowHelper.ThrowObjectDisposedException(nameof(OwnedMemory<T>), ExceptionResource.Memory_ThrowIfDisposed);
+                return new ReadOnlyMemory<T>(this, 0, Length);
+            }
+        }
```

Comment:
Why do we need `ReadOnlyMemory`?  We expose `Memory`, which is efficiently convertible to `ReadOnlyMemory`, right?  It just seems inconsistent when, for example, we don't expose both `AsSpan` and `AsReadOnlySpan`, just `AsSpan`.

Also, from a naming perspective, why is it `AsSpan` but just `Memory` rather than having `Span`/`Memory` or `AsSpan`/`AsMemory`?